Application freezes when injecting DataSource bean into GORM data service

### Task List

- [x] Steps to reproduce provided
- [ ] Stacktrace (if present) provided
- [x] Example that reproduces the problem uploaded to Github
- [x] Full description of the issue provided (see below)

### Steps to Reproduce

1. Create a factory bean for the purpose of creating a`groovy.sql.Sql` bean
2. Create a GORM data service that has said bean injected
3. Run anything that causes the data service bean to be instantiated

### Expected Behaviour

A DataSource bean is injected into the Sql factory method

### Actual Behaviour

The application freezes when retrieving the bean from the ApplicationContext 

### Environment Information

- **Operating System**: MacOS 10.15.7
- **Micronaut Version:** 2.3.1
- **JDK Version:** 11

### Example Application

- https://github.com/jamesdh/demo/tree/micronaut_groovy_185

Coming back to this, if I try to inject *anything* that comes from HibernateDatastoreFactory (HibernateDatastore, SessionFactory, DataSource, or PlatformTransactionManager), I get an excruciatingly long pause, eventually followed by a flurry of the following for about a minute:

    [Test worker] DEBUG 🐳 [postgres:12] - Trying to create JDBC connection using org.postgresql.Driver to jdbc:postgresql://localhost:55345/tt?loggerLevel=OFF with properties: {password=test, user=test}

...eventually followed by: 

```
SEVERE: Unable to create initial connections of pool.
java.sql.SQLException: Could not create new connection
	at org.testcontainers.containers.JdbcDatabaseContainer.createConnection(JdbcDatabaseContainer.java:228)
	at org.testcontainers.jdbc.ContainerDatabaseDriver.connect(ContainerDatabaseDriver.java:124)
	at org.apache.tomcat.jdbc.pool.PooledConnection.connectUsingDriver(PooledConnection.java:319)
	at org.apache.tomcat.jdbc.pool.PooledConnection.connect(PooledConnection.java:212)
	at org.apache.tomcat.jdbc.pool.ConnectionPool.createConnection(ConnectionPool.java:744)
	at org.apache.tomcat.jdbc.pool.ConnectionPool.borrowConnection(ConnectionPool.java:676)
	at org.apache.tomcat.jdbc.pool.ConnectionPool.init(ConnectionPool.java:483)
	at org.apache.tomcat.jdbc.pool.ConnectionPool.<init>(ConnectionPool.java:154)
	at org.apache.tomcat.jdbc.pool.DataSourceProxy.pCreatePool(DataSourceProxy.java:118)
	at org.apache.tomcat.jdbc.pool.DataSourceProxy.createPool(DataSourceProxy.java:107)
	at org.apache.tomcat.jdbc.pool.DataSourceProxy.getConnection(DataSourceProxy.java:131)
	at org.springframework.jdbc.datasource.LazyConnectionDataSourceProxy$LazyConnectionInvocationHandler.getTargetConnection(LazyConnectionDataSourceProxy.java:405)
	at org.springframework.jdbc.datasource.LazyConnectionDataSourceProxy$LazyConnectionInvocationHandler.invoke(LazyConnectionDataSourceProxy.java:378)
	at com.sun.proxy.$Proxy105.getAutoCommit(Unknown Source)
	at jdk.internal.reflect.GeneratedMethodAccessor132.invoke(Unknown Source)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.springframework.jdbc.datasource.TransactionAwareDataSourceProxy$TransactionAwareInvocationHandler.invoke(TransactionAwareDataSourceProxy.java:238)
	at com.sun.proxy.$Proxy105.getAutoCommit(Unknown Source)
	at org.hibernate.resource.transaction.backend.jdbc.internal.DdlTransactionIsolatorNonJtaImpl.getIsolatedConnection(DdlTransactionIsolatorNonJtaImpl.java:46)
	at org.hibernate.tool.schema.internal.exec.ImprovedExtractionContextImpl.getJdbcConnection(ImprovedExtractionContextImpl.java:60)
	at org.hibernate.tool.schema.extract.internal.SequenceInformationExtractorLegacyImpl.extractMetadata(SequenceInformationExtractorLegacyImpl.java:40)
	at org.hibernate.tool.schema.extract.internal.DatabaseInformationImpl.initializeSequences(DatabaseInformationImpl.java:65)
	at org.hibernate.tool.schema.extract.internal.DatabaseInformationImpl.<init>(DatabaseInformationImpl.java:59)
	at org.hibernate.tool.schema.internal.Helper.buildDatabaseInformation(Helper.java:155)
	at org.hibernate.tool.schema.internal.AbstractSchemaMigrator.doMigration(AbstractSchemaMigrator.java:96)
	at org.hibernate.tool.schema.spi.SchemaManagementToolCoordinator.performDatabaseAction(SchemaManagementToolCoordinator.java:184)
	at org.hibernate.tool.schema.spi.SchemaManagementToolCoordinator.process(SchemaManagementToolCoordinator.java:73)
	at org.hibernate.internal.SessionFactoryImpl.<init>(SessionFactoryImpl.java:316)
	at org.hibernate.boot.internal.SessionFactoryBuilderImpl.build(SessionFactoryBuilderImpl.java:469)
	at org.hibernate.cfg.Configuration.buildSessionFactory(Configuration.java:708)
	at org.grails.orm.hibernate.cfg.HibernateMappingContextConfiguration.buildSessionFactory(HibernateMappingContextConfiguration.java:287)
	at org.grails.orm.hibernate.connections.HibernateConnectionSourceFactory.create(HibernateConnectionSourceFactory.java:86)
	at org.grails.orm.hibernate.connections.AbstractHibernateConnectionSourceFactory.create(AbstractHibernateConnectionSourceFactory.java:39)
	at org.grails.orm.hibernate.connections.AbstractHibernateConnectionSourceFactory.create(AbstractHibernateConnectionSourceFactory.java:23)
	at org.grails.datastore.mapping.core.connections.AbstractConnectionSourceFactory.create(AbstractConnectionSourceFactory.java:64)
	at org.grails.datastore.mapping.core.connections.AbstractConnectionSourceFactory.create(AbstractConnectionSourceFactory.java:52)
	at org.grails.datastore.mapping.core.connections.ConnectionSourcesInitializer.create(ConnectionSourcesInitializer.groovy:24)
	at org.grails.orm.hibernate.HibernateDatastore.<init>(HibernateDatastore.java:212)
	at org.grails.orm.hibernate.HibernateDatastore.<init>(HibernateDatastore.java:233)
	at io.micronaut.configuration.hibernate.gorm.HibernateDatastoreFactory.hibernateDatastore(HibernateDatastoreFactory.groovy:63)
	at io.micronaut.configuration.hibernate.gorm.$HibernateDatastoreFactory$HibernateDatastore0Definition.build(Unknown Source)
	at io.micronaut.context.DefaultBeanContext.doCreateBean(DefaultBeanContext.java:1898)
	at io.micronaut.context.DefaultBeanContext.createAndRegisterSingletonInternal(DefaultBeanContext.java:2679)
	at io.micronaut.context.DefaultBeanContext.loadContextScopeBean(DefaultBeanContext.java:2215)
	at io.micronaut.context.DefaultBeanContext.initializeContext(DefaultBeanContext.java:1543)
	at io.micronaut.context.DefaultApplicationContext.initializeContext(DefaultApplicationContext.java:220)
	at io.micronaut.context.DefaultBeanContext.readAllBeanDefinitionClasses(DefaultBeanContext.java:2812)
	at io.micronaut.context.DefaultBeanContext.start(DefaultBeanContext.java:233)
	...truncated
Caused by: org.postgresql.util.PSQLException: FATAL: sorry, too many clients already
	at org.postgresql.core.v3.ConnectionFactoryImpl.doAuthentication(ConnectionFactoryImpl.java:613)
	at org.postgresql.core.v3.ConnectionFactoryImpl.tryConnect(ConnectionFactoryImpl.java:161)
	at org.postgresql.core.v3.ConnectionFactoryImpl.openConnectionImpl(ConnectionFactoryImpl.java:213)
	at org.postgresql.core.ConnectionFactory.openConnection(ConnectionFactory.java:51)
	at org.postgresql.jdbc.PgConnection.<init>(PgConnection.java:225)
	at org.postgresql.Driver.makeConnection(Driver.java:465)
	at org.postgresql.Driver.connect(Driver.java:264)
	at org.testcontainers.containers.JdbcDatabaseContainer.createConnection(JdbcDatabaseContainer.java:219)
	... 124 more
```

and then finally 

```
[Test worker] DEBUG o.h.e.jdbc.spi.SqlExceptionHelper - Unable to check JDBC Connection auto-commit in preparation for DDL execution [n/a]
java.sql.SQLException: Could not create new connection
```

--- a/hibernate-gorm/src/main/groovy/io/micronaut/configuration/hibernate/gorm/HibernateDatastoreFactory.groovy
+++ b/hibernate-gorm/src/main/groovy/io/micronaut/configuration/hibernate/gorm/HibernateDatastoreFactory.groovy
@@ -32,6 +32,8 @@ import org.grails.orm.hibernate.connections.HibernateConnectionSource
 import org.hibernate.SessionFactory
 import org.springframework.transaction.PlatformTransactionManager
 
+import javax.inject.Named
+import javax.inject.Singleton
 import javax.sql.DataSource
 import java.util.stream.Stream
 
@@ -85,5 +87,23 @@ class HibernateDatastoreFactory {
 
         return datastore
     }
+
+    @Singleton
+    SessionFactory sessionFactory(HibernateDatastore hibernateDatastore) {
+        hibernateDatastore.getSessionFactory()
+    }
+
+    @Singleton
+    DataSource dataSource(HibernateDatastore hibernateDatastore) {
+        ((HibernateConnectionSource) hibernateDatastore.getConnectionSources().defaultConnectionSource).getDataSource()
+    }
+
+    @Primary
+    @Named("hibernate")
+    @Singleton
+    PlatformTransactionManager transactionManager(HibernateDatastore hibernateDatastore) {
+        hibernateDatastore.getTransactionManager()
+    }
+
 
 }

--- a/hibernate-gorm/src/main/groovy/io/micronaut/configuration/hibernate/gorm/HibernateDatastoreFactory.groovy
+++ b/hibernate-gorm/src/main/groovy/io/micronaut/configuration/hibernate/gorm/HibernateDatastoreFactory.groovy
@@ -25,13 +25,12 @@ import io.micronaut.context.annotation.Bean
 import io.micronaut.context.annotation.Context
 import io.micronaut.context.annotation.Factory
 import io.micronaut.context.annotation.Primary
+import io.micronaut.inject.qualifiers.Qualifiers
 import org.grails.orm.hibernate.HibernateDatastore
 import org.grails.orm.hibernate.connections.HibernateConnectionSource
 import org.hibernate.SessionFactory
 import org.springframework.transaction.PlatformTransactionManager
 
-import javax.inject.Named
-import javax.inject.Singleton
 import javax.sql.DataSource
 import java.util.stream.Stream
 
@@ -63,33 +62,20 @@ class HibernateDatastoreFactory {
             new ConfigurableEventPublisherAdapter(applicationContext),
             classes
         )
+        SessionFactory sessionFactory = datastore.sessionFactory
+        DataSource dataSource = ((HibernateConnectionSource) datastore.getConnectionSources().defaultConnectionSource).getDataSource()
+        PlatformTransactionManager platformTransactionManager = datastore.getTransactionManager()
+        applicationContext.registerSingleton(SessionFactory, sessionFactory)
+        applicationContext.registerSingleton(DataSource, dataSource)
+        applicationContext.registerSingleton(PlatformTransactionManager, platformTransactionManager, Qualifiers.byName("hibernate"))
+        applicationContext.registerSingleton(PlatformTransactionManager, platformTransactionManager, Qualifiers.byStereotype(Primary))
+        for (o in datastore.getServices()) {
+            applicationContext.registerSingleton(o, false)
+        }
+        for (o in datastore.getServices()) {
+            applicationContext.inject(o)
+        }
         return datastore
     }
 
-    @Singleton
-    SessionFactory sessionFactory(HibernateDatastore hibernateDatastore) {
-        hibernateDatastore.getSessionFactory()
-    }
-
-    @Singleton
-    DataSource dataSource(HibernateDatastore hibernateDatastore) {
-        ((HibernateConnectionSource) hibernateDatastore.getConnectionSources().defaultConnectionSource).getDataSource()
-    }
-
-    @Primary
-    @Named("hibernate")
-    @Singleton
-    PlatformTransactionManager transactionManager(HibernateDatastore hibernateDatastore) {
-        hibernateDatastore.getTransactionManager()
-    }
-
-    @Singleton
-    void setupServices(HibernateDatastore hibernateDatastore, SessionFactory sessionFactory, DataSource dataSource, PlatformTransactionManager platformTransactionManager) {
-        for (o in hibernateDatastore.getServices()) {
-            applicationContext.registerSingleton(o, false)
-        }
-        for (o in hibernateDatastore.getServices()) {
-            applicationContext.inject(o)
-        }
-    }
 }

--- a/hibernate-gorm/src/main/groovy/io/micronaut/configuration/hibernate/gorm/HibernateDatastoreFactory.groovy
+++ b/hibernate-gorm/src/main/groovy/io/micronaut/configuration/hibernate/gorm/HibernateDatastoreFactory.groovy
@@ -26,6 +26,7 @@ import io.micronaut.context.annotation.Context
 import io.micronaut.context.annotation.Factory
 import io.micronaut.context.annotation.Primary
 import io.micronaut.inject.qualifiers.Qualifiers
+import org.grails.datastore.mapping.services.Service
 import org.grails.orm.hibernate.HibernateDatastore
 import org.grails.orm.hibernate.connections.HibernateConnectionSource
 import org.hibernate.SessionFactory
@@ -69,12 +70,19 @@ class HibernateDatastoreFactory {
         applicationContext.registerSingleton(DataSource, dataSource)
         applicationContext.registerSingleton(PlatformTransactionManager, platformTransactionManager, Qualifiers.byName("hibernate"))
         applicationContext.registerSingleton(PlatformTransactionManager, platformTransactionManager, Qualifiers.byStereotype(Primary))
-        for (o in datastore.getServices()) {
-            applicationContext.registerSingleton(o, false)
+
+        Collection<Collection<Service>> gormServices = datastore.getServices().split {
+            !it.class.isAnnotationPresent(grails.gorm.services.Service)
         }
-        for (o in datastore.getServices()) {
-            applicationContext.inject(o)
+        for(services in gormServices) {
+            for (o in services) {
+                applicationContext.registerSingleton(o, false)
+            }
+            for (o in services) {
+                applicationContext.inject(o)
+            }
         }
+
         return datastore
     }
 

--- a/hibernate-gorm/src/main/groovy/io/micronaut/configuration/hibernate/gorm/HibernateDatastoreFactory.groovy
+++ b/hibernate-gorm/src/main/groovy/io/micronaut/configuration/hibernate/gorm/HibernateDatastoreFactory.groovy
@@ -63,13 +63,6 @@ class HibernateDatastoreFactory {
             new ConfigurableEventPublisherAdapter(applicationContext),
             classes
         )
-        for (o in datastore.getServices()) {
-            applicationContext.registerSingleton(o, false)
-        }
-        for (o in datastore.getServices()) {
-            applicationContext.inject(o)
-        }
-
         return datastore
     }
 
@@ -88,5 +81,15 @@ class HibernateDatastoreFactory {
     @Singleton
     PlatformTransactionManager transactionManager(HibernateDatastore hibernateDatastore) {
         hibernateDatastore.getTransactionManager()
+    }
+
+    @Singleton
+    void setupServices(HibernateDatastore hibernateDatastore, SessionFactory sessionFactory, DataSource dataSource, PlatformTransactionManager platformTransactionManager) {
+        for (o in hibernateDatastore.getServices()) {
+            applicationContext.registerSingleton(o, false)
+        }
+        for (o in hibernateDatastore.getServices()) {
+            applicationContext.inject(o)
+        }
     }
 }

--- a/hibernate-gorm/src/test/groovy/io/micronaut/configuration/hibernate/gorm/GormConfigSpec.groovy
+++ b/hibernate-gorm/src/test/groovy/io/micronaut/configuration/hibernate/gorm/GormConfigSpec.groovy
@@ -216,6 +216,7 @@ class Book {
 abstract class BookService {
 
     @Inject DataSource dataSource
+    @Inject TransactionService transactionService
 
     @Value('${data-source.db-create}')
     String dbCreate

--- a/hibernate-gorm/src/test/groovy/io/micronaut/configuration/hibernate/gorm/GormConfigSpec.groovy
+++ b/hibernate-gorm/src/test/groovy/io/micronaut/configuration/hibernate/gorm/GormConfigSpec.groovy
@@ -18,6 +18,7 @@ package io.micronaut.configuration.hibernate.gorm
 import grails.gorm.annotation.Entity
 import grails.gorm.services.Service
 import grails.gorm.transactions.TransactionService
+import io.micronaut.context.annotation.Primary
 import io.micronaut.inject.qualifiers.Qualifiers
 import org.grails.datastore.mapping.validation.ValidationException
 import org.grails.orm.hibernate.GrailsHibernateTransactionManager
@@ -31,6 +32,7 @@ import spock.lang.Shared
 import spock.lang.Specification
 
 import javax.annotation.PostConstruct
+import javax.inject.Inject
 import javax.inject.Singleton
 import javax.sql.DataSource
 
@@ -56,6 +58,9 @@ class GormConfigSpec extends Specification {
 
         and:
         applicationContext.containsBean(PlatformTransactionManager, Qualifiers.byName("hibernate"))
+
+        and:
+        applicationContext.containsBean(PlatformTransactionManager, Qualifiers.byStereotype(Primary))
 
         and: 'two beans: one created by the factory and the MockPlatformTransactionManager'
         applicationContext.getBeansOfType(PlatformTransactionManager).size() == 2
@@ -209,6 +214,8 @@ class Book {
 @Service(Book)
 @Singleton
 abstract class BookService {
+
+    @Inject DataSource dataSource
 
     @Value('${data-source.db-create}')
     String dbCreate

